src/vos/vos_aggregate.c

```diff
@@ -1212,24 +1278,45 @@ insert_segments(daos_handle_t ih, struct agg_merge_window *mw,
 			goto abort;
 		}
 
-		/* Physical entry is in window */
-		if (rect.rc_ex.ex_hi <= mw->mw_ext.ex_hi) {
+		/* Physical entry is in window or fully removed */
+		if (rect.rc_ex.ex_hi <= mw->mw_ext.ex_hi ||
+		    phy_ent->pe_remove) {
 			d_list_del(&phy_ent->pe_link);
-			D_FREE_PTR(phy_ent);
+			D_FREE(phy_ent);
 			D_ASSERT(mw->mw_phy_cnt > 0);
 			mw->mw_phy_cnt--;
 			continue;
 		}
-
 		/* Update extent start of truncated physical entry */
 		rect.rc_ex.ex_lo = mw->mw_ext.ex_hi + 1;
 		phy_ent->pe_off = rect.rc_ex.ex_lo -
 				phy_ent->pe_rect.rc_ex.ex_lo;
 		phy_ent->pe_trunc_head = false;
+
 		leftovers++;
 	}
 	D_ASSERT(leftovers == mw->mw_phy_cnt);
 
+	/** Remove processed removal records */
+	d_list_for_each_entry_safe(rm_ent, rm_tmp, &mw->mw_rmv_ents, re_link) {
+		rect = rm_ent->re_rect;
+
+		if (!last && rect.rc_ex.ex_hi > mw->mw_ext.ex_hi)
```

Comment:
I'm not sure why 'last' needs be checked here, we guaranteed last window covers remaining 'deleted' records, no?

Reply:
Merge window is for visible extents.   Suppose you have a physical extent from 0-10 and it's covered by a removal record from 1-10.   The window will be flushed when we process the delete record which has "last" set but in this case, the delete record has hi > mw_ext.ex_hi so it will not get deleted unless we pass the "last" flag too

Reply:
indeed.